In [2]:
import pandas as pd

## 1️⃣ Data Loading

In [3]:
file_dir = "../data/raw/raw-titanic-dataset.csv"

df = pd.read_csv(file_dir)

In [4]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [5]:
print("The dataset columns are:", end="\n")
df.columns

The dataset columns are:


Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

### This dataset contains 891 passengers and 12 features

In [6]:
print(f"This is the shape of the Dataframe: {df.shape}")

This is the shape of the Dataframe: (891, 12)


In [7]:
print("List of dataset data types: ", end="\n")
df.dtypes

List of dataset data types: 


PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

In [8]:
missing_values = df.isnull().sum()
print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


## 2️⃣ Data Cleaning

### There are some datas that have null values:
- Age: 177 rows
- Cabin: 687 rows
- Embarked: 2 rows

### This is the next steps:
1. Replace the missing values of age column with the median of the ages
2. Replace the missing values of embarked column with mode method
3. Drop `Cabin` & `Ticket` columns
4. Verify the dataset has no more `NaN` values & the columns are correct
5. Save the cleaned dataset

### 1️⃣ Filling Missing Values in Age Using Median
The Age column contains missing values that must be handled before performing analysis.

Since Age is a numerical variable, we use the median to fill missing values. The median is preferred over the mean because it is less affected by outliers and skewed distributions.

By using the median:
- We maintain a realistic central value.
- We reduce the impact of extreme age values.
- We preserve the overall distribution of the dataset.

After filling the missing values, we verify the results by checking the remaining null values in the dataset.

In [9]:
median_age = df['Age'].median()
df.fillna({'Age': median_age }, inplace=True)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [10]:
missing_values = df.isnull().sum()
print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


### 2️⃣ Filling Missing Values in Embarked Using Mode
The Embarked column contains missing values that need to be handled before further analysis.

Since Embarked is a categorical variable, it is appropriate to fill missing values using the mode, which represents the most frequently occurring category in the dataset.

Using the mode ensures that:
- We preserve the distribution of the most common embarkation port.
- We avoid introducing unrealistic or biased values.
- The dataset remains consistent for categorical analysis.

After applying the imputation, we verify that the missing values have been successfully replaced by checking the null value count again.

In [11]:
frq_embarked = df['Embarked'].mode()[0]
df.fillna({'Embarked': frq_embarked }, inplace=True)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female  28.0      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

as you see, we have to add `[0]` to the mode method to get the value. This because the mode() doesn't return `scalar` value, but `series`, so we have to call the first row value.

In [12]:
missing_values = df.isnull().sum()
print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64


### 3️⃣ Drop Unused Columns
In data preprocessing, it is important to remove irrelevant or low-value columns to simplify analysis and reduce noise.

For this dataset, we drop the following columns:
- Cabin
- Ticket

#### 🔎 Why Drop These Columns?
- **Cabin**
  - Contains a large number of missing values (687 null entries).
  - Highly sparse and difficult to impute reliably.
  - High cardinality with limited predictive value in this stage.
- **Ticket**
  - Mostly unique values.
  - High cardinality feature.
  - Difficult to extract meaningful patterns without additional feature engineering.


In [13]:
df.drop(columns=['Cabin', 'Ticket'], inplace=True)
missing_values = df.isnull().sum()
print(missing_values)

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64


### 4️⃣ Verify Columns and Data Integrity
After cleaning and transformation, we must verify that:
- All required columns exist.
- No missing values remain.
- Data types are correct.
- The dataset structure is ready for modeling.

In [14]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Fare', 'Embarked'],
      dtype='str')

In [15]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64

In [16]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Fare         891 non-null    float64
 9   Embarked     891 non-null    str    
dtypes: float64(2), int64(5), str(3)
memory usage: 69.7 KB


### 5️⃣ Save the Cleaned Dataset
The final step is to export the cleaned dataset into a new file for future use.

We save the dataset in CSV format.

In [17]:
# uncomment the code below to save the cleaned titanic dataset

# df.to_csv("../data/processed/cleaned-titanic-dataset.csv", index=False, encoding='utf-8')

## 3️⃣ Exploratory Data Analysis (EDA)

### 👥 3.1 Survival Rate by Gender

In [18]:
df.groupby(['Sex']).size()

Sex
female    314
male      577
dtype: int64

### Interpretetion

From the data above we can see the numbers of passengers grouped by their gender
- There are 577 male passengers onboard
- There are 314 female passengers onboard

We can see that _**64.76%**_ of passengers were male, and _**35.24%**_ were female.

In [19]:
df.groupby(['Sex', 'Survived']).size()

Sex     Survived
female  0            81
        1           233
male    0           468
        1           109
dtype: int64

### Interpretation
From the grouped data, we observe:
- There were 314 female passengers and 577 male passengers onboard.
- 233 females survived, while only 109 males survived.

When calculating survival rates within each gender:
- Female survival rate ≈ 74%
- Male survival rate ≈ 19%

This shows a significant difference in survival probability between genders. Female passengers were nearly four times more likely to survive than male passengers. This strongly supports the historical “women and children first” evacuation policy during the Titanic disaster.

### 🛳️ 3.2 Survival Rate by Passenger's Class

In [20]:
df.groupby(['Pclass', 'Survived']).size()

Pclass  Survived
1       0            80
        1           136
2       0            97
        1            87
3       0           372
        1           119
dtype: int64

### Interpretation
From the grouped data, we observe:
- There were:
  - **216 passengers** on the class 1 onboard
  - **184 passengers** on the class 2 onboard
  - **491 passengers** on the class 3 onboard

When calculating survival rates within each passenger class:
- Class 1 survival rate ≈ 63%
- Class 2 survival rate ≈ 47%
- Class 3 survival rate ≈ 24%

There is a clear downward trend in survival probability as passenger class decreases. First-class passengers were more than twice as likely to survive compared to third-class passengers.

This suggests a strong association between socioeconomic status (represented by passenger class) and survival outcomes. The disparity may reflect differences in cabin location, access to lifeboats, or evacuation prioritization during the disaster.

### 🧔🏻‍♂️ 3.3.1 Survival Rate by Passenger's Age

In [21]:
df.groupby('Survived').agg(median_age=('Age', 'median'), mean_age=('Age', 'mean'))

median_age   mean_age
Survived                       
0               28.0  30.028233
1               28.0  28.291433

### Interpretation
From the grouped data, we observe:
- The median age of both survivors and non-survivors is identical at 28 years.
- The mean age differs slightly by approximately 1.7 years, with non-survivors being marginally older on average.

The small difference in mean age and identical median values suggest that age alone does not have a strong relationship with survival outcomes. Based on this aggregate analysis, age does not appear to be a strong standalone factor influencing survival probability.

### 👦🏻 3.3.2 Children Survival Rate

In [22]:
df['Age Group'] = pd.cut(
    df['Age'],
    bins=[0,14,24,64,120],
    labels=['children','youth','adult','elderly']
)

### Interpretation
Based on common demographic and data grouping, we can seperate the age group as:
- 0-14: children
- 15-24: youth
- 25-64: adult
- 65+: elderly

In [25]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch     Fare Embarked Age Group  
0      0   7.2500        S     youth  
1      0  71.2833        C     adult  
2      0   7.9250        S     adult  
3      0  53.1000        S     adult  
4      0   8.0500        S     adult

In [26]:
df.groupby(['Age Group', 'Survived']).size()

Age Group  Survived
adult      0           379
           1           223
children   0            32
           1            45
elderly    0            10
           1             1
youth      0           126
           1            73
dtype: int64

From the grouped data, we observe:
- There were:
  - **45 children** were survived from 77 children
  - **73 youths** were survived from 199 youth
  - **223 adults** were survived from 602 adults
  - **1 elder** were survived from 11 elders

When calculating survival rates within each passenger age group:
- Children survival rate ≈ 58%
- Youth survival rate ≈ 37%
- Adult survival rate ≈ 37%
- Elderly survival rate ≈ 9%

There is a clear downward trend in survival probability from children to elderly.

The higher survival rate among children is consistent with the historical evacuation practice prioritizing women and children.


### 👦🏻 3.3.3 Age Group + Gender Survival Rate

In [28]:
df.groupby(['Age Group', 'Sex', 'Survived']).size()

Age Group  Sex     Survived
adult      female  0            49
                   1           148
           male    0           330
                   1            75
children   female  0            14
                   1            24
           male    0            18
                   1            21
elderly    male    0            10
                   1             1
youth      female  0            17
                   1            61
           male    0           109
                   1            12
dtype: int64

### Interpretation
From the grouped data, we observe:
- There were:
  - **148 female adults** were survived from 197 female adults
  - **75 male adults** were survived from 405 male adults

When calculating survival rates within each gender on adult passengers:
- Female adults survival rate ≈ 75%
- Male adults survival rate ≈ 19%

There is a substantial survival gap between adult females and adult males.

The higher survival rate among female adults is also consistent with the historical evacuation practice prioritizing women and children.


### 🛳️ 3.4 Survival Rate by Embarkation Port

In [23]:
df.groupby(['Embarked', 'Survived']).size()

Embarked  Survived
C         0            75
          1            93
Q         0            47
          1            30
S         0           427
          1           219
dtype: int64

### Interpretation
From the grouped data, we observe:
- There were:
  - **168 passengers** embarked from port C which has 93 survivors
  - **77 passengers** embarked from port Q which has 30 survivors
  - **646 passengers** embarked from port S which has 219 survivors

When calculating survival rates within each embarkation port:
- Port C survival rate ≈ 55%
- Port Q survival rate ≈ 39%
- Port S survival rate ≈ 34%

There is a clear downward trend in survival probability between each port. Although the embarkation port itself can't be a solid prove of survival probability. So, we need the next step, which grouping the embarkation port and passenger's class

In [24]:
df.groupby(['Embarked', 'Pclass', 'Survived']).size()

Embarked  Pclass  Survived
C         1       0            26
                  1            59
          2       0             8
                  1             9
          3       0            41
                  1            25
Q         1       0             1
                  1             1
          2       0             1
                  1             2
          3       0            45
                  1            27
S         1       0            53
                  1            76
          2       0            88
                  1            76
          3       0           286
                  1            67
dtype: int64

From the grouped data, we observe:
- There were:
  - Port C:
    - **59 survivors** from 85 class 1 passengers
    - **9 survivors** from 17 class 2 passengers
    - **25 survivors** from 66 class 3 passengers
  - Port Q:
    - **1 survivors** from 2 class 1 passengers
    - **2 survivors** from 3 class 2 passengers
    - **27 survivors** from 72 class 3 passengers
  - Port S:
    - **76 survivors** from 129 class 1 passengers
    - **76 survivors** from 164 class 2 passengers
    - **67 survivors** from 353 class 3 passengers

When calculating passenger class rate within each embarkation port:
- Port C:
  - Class 1 ≈ 51%
  - Class 2 ≈ 10%
  - Class 3 ≈ 39%
- Port Q:
  - Class 1 ≈ 3%
  - Class 2 ≈ 4%
  - Class 3 ≈ 93%
- Port S:
  - Class 1 ≈ 20%
  - Class 2 ≈ 25%
  - Class 3 ≈ 55%

This suggests that the higher survival rate at Port C may be influenced by a larger proportion of first-class passengers. Although there was more class 1 passengers who went through Port S, but the Port S was dominated by the class 3 passengers.

Most of the class 1 passengers were directed through Port C & S. And from the embarked data earlier, we can see most of all passengers went through Port S.